codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttEncoder.java

```diff
@@ -689,10 +697,33 @@ private static void writeVariableLengthInt(ByteBuf buf, int num) {
         } while (num > 0);
     }
 
-    static void writeUTF8String(ByteBuf buf, String s) {
-        byte[] sBytes = encodeStringUtf8(s);
-        buf.writeShort(sBytes.length);
-        buf.writeBytes(sBytes, 0, sBytes.length);
+    private static int nullableUtf8Bytes(String s) {
+        if (s == null) {
+            return 0;
+        }
+        return utf8Bytes(s);
+    }
+
+    private static void writeUTF8String(ByteBuf buf, String s, int expectedUtf8Length) {
+        assert nullableUtf8Bytes(s) == expectedUtf8Length;
+        assert buf.isWritable(expectedUtf8Length + 2);
+        buf.writeShort(expectedUtf8Length);
+        if (expectedUtf8Length > 0) {
+            final int utf8Length = reserveAndWriteUtf8(buf, s, expectedUtf8Length);
+            assert expectedUtf8Length == utf8Length;
+        }
+    }
+
+    private static void writeUTF8String(ByteBuf buf, String s) {
+        assert buf.isWritable(nullableUtf8Bytes(s) + 2);
+        final int writerIndex = buf.writerIndex();
+        buf.ensureWritable(buf.capacity() - writerIndex);
```

Comment:
This is redundant for the same reason, but this whole method should only be used when it's already known that the buffer has sufficient space.. maybe rename it `writeUTF8StringUnsafe` or something

Reply:
> this whole method should only be used when it's already known that the buffer has sufficient space

Agree as I've said on https://github.com/netty/netty/pull/10509#discussion_r481810148:

> I would like the whole required writable capacity to be already allocated and the assert should just check it.

So, how to do it?
As you've pointed on https://github.com/netty/netty/pull/10509#discussion_r481458743 
> it won't affect how the buffer is allocated

So, the only reasonable way to ensure there is enough writable space is to call `ensureWritable` right after allocating the whole buffer, once. wdyt?
And we can save `buf.ensureWritable(buf.capacity() - writerIndex);` to be called on `writeUTF8String`

Reply:
I've changed it by assuming `Or it isn't needed at all if the initialCapacity of the allocated buffer is enough?` , please take a look and thanks for the detailed review Nick :)